src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.diff.sql

```diff
@@ -83,5 +83,229 @@ AS
+-- STORED PROCEDURE
+--    Reindex instance
+--
+-- DESCRIPTION
+--    Reidex instance
+--
+-- PARAMETERS
+--     @studyInstanceUid
+--         * The study instance UID.
+--     @seriesInstanceUid
+--         * The series instance UID.
+--     @sopInstanceUid
+--         * The SOP instance UID.
+--     @stringExtendedQueryTags
+--         * String extended query tag data
+--     @longExtendedQueryTags
+--         * Long extended query tag data
+--     @doubleExtendedQueryTags
+--         * Double extended query tag data
+--     @dateTimeExtendedQueryTags
+--         * DateTime extended query tag data
+--     @personNameExtendedQueryTags
+--         * PersonName extended query tag data
+
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.ReindexInstance
+    @studyInstanceUid       VARCHAR(64),
+    @seriesInstanceUid      VARCHAR(64),
+    @sopInstanceUid         VARCHAR(64),               
+    @stringExtendedQueryTags dbo.InsertStringExtendedQueryTagTableType_1 READONLY,    
+    @longExtendedQueryTags dbo.InsertLongExtendedQueryTagTableType_1 READONLY,
+    @doubleExtendedQueryTags dbo.InsertDoubleExtendedQueryTagTableType_1 READONLY,
+    @dateTimeExtendedQueryTags dbo.InsertDateTimeExtendedQueryTagTableType_1 READONLY,
+    @personNameExtendedQueryTags dbo.InsertPersonNameExtendedQueryTagTableType_1 READONLY
+AS
+    SET NOCOUNT    ON
+    SET XACT_ABORT ON
+    BEGIN TRANSACTION
+
+        DECLARE @studyKey BIGINT
+        DECLARE @seriesKey BIGINT
+        DECLARE @instanceKey BIGINT
+        DECLARE @watermark BIGINT
+
+        -- Add lock so that the instance won't be removed
+        DECLARE @status TINYINT
+        SELECT
+            @studyKey = StudyKey,
+            @seriesKey = SeriesKey,
+            @instanceKey = InstanceKey,
+            @watermark = Watermark,
+            @status = Status
+        FROM dbo.Instance WITH (HOLDLOCK) 
+        WHERE StudyInstanceUid = @studyInstanceUid
+            AND SeriesInstanceUid = @seriesInstanceUid
+            AND SopInstanceUid = @sopInstanceUid
+
+        IF @@ROWCOUNT = 0
+            THROW 50404, 'Instance not exists', 1
+        IF @status <> 1 -- Created
+            THROW 50409, 'Instance is not been stored succssfully', 1
+
+        -- Insert Extended Query Tags
+
+        -- String Key tags
+        IF EXISTS (SELECT 1 FROM @stringExtendedQueryTags)
+        BEGIN      
+            MERGE INTO dbo.ExtendedQueryTagString AS T
+            USING 
+            (
+                -- Locks tags in dbo.ExtendedQueryTag
+                SELECT input.TagKey, input.TagValue, input.TagLevel 
+                FROM @stringExtendedQueryTags input
+                INNER JOIN dbo.ExtendedQueryTag WITH (REPEATABLEREAD) 
+                ON dbo.ExtendedQueryTag.TagKey = input.TagKey
+                -- Not merge on extended query tag which is being deleted.
+                AND dbo.ExtendedQueryTag.TagStatus <> 2     
```

Comment:
Would it be better if we:
1. Only operate on tags that are currently `Adding` (1)
2. The `OperationId` is the expected value

Reply:
Regarding 2:   is there a scenario we have to check `OperationId`? The storage procedure is really long now, trying to simplify and make it as short as possible..

Reply:
Based on our conversation, we can skip (2). However, we also need to make sure the reindex procedure can run multiple times without issue. Looking at the implementation, it seems fine though